lib/rig_inbound_gateway/api_proxy/response_from_parser.ex

```diff
@@ -0,0 +1,103 @@
+defmodule RigInboundGateway.ApiProxy.ResponseFromParser do
+  @moduledoc """
+  Handles parsing of messages coming from different sources (e.g. Kafka)
+
+  """
+
+  require Logger
+
+  use Rig.Config, [:schema_registry_host]
+
+  alias Rig.Connection.Codec
+  alias RigKafka.Avro
+
+  @default_response_code 200
+
+  # ---
+
+  @spec parse([{String.t(), String.t()}, ...], map()) ::
+          {pid, pos_integer(), any(), map()} | String.t()
+  # structured
+  def parse(
+        _headers,
+        %{
+          "body" => body,
+          "rig" => rig_metadata
+        } = message
+      ) do
+    with {:ok, correlation_id} <- Map.fetch(rig_metadata, "correlation"),
+         {:ok, deserialized_pid} <- Codec.deserialize(correlation_id),
+         raw_response_code <- Map.get(rig_metadata, "response_code", @default_response_code),
+         # convert status code to int if needed, HTTP headers can't have number as a value
+         {response_code, _} <- to_int(raw_response_code),
+         response_headers <- Map.get(message, "headers", %{}),
+         response_body <- try_encode(body) do
+      Logger.debug(fn ->
+        "Parsed structured HTTP response: body=#{inspect(response_body)}, headers=#{
+          inspect(response_headers)
+        }, code=#{inspect(response_code)}"
+      end)
+
+      {deserialized_pid, response_code, response_body, response_headers}
+    else
+      err -> err
+    end
+  end
+
+  # binary
+  def parse(
+        headers,
+        message
+      ) do
+    with headers_map <- Enum.into(headers, %{}),
+         {:ok, correlation_id} <- Map.fetch(headers_map, "rig-correlation"),
+         {:ok, deserialized_pid} <- Codec.deserialize(correlation_id),
+         raw_response_code <- Map.get(headers_map, "rig-response-code", @default_response_code),
+         # convert status code to int if needed, HTTP headers can't have number as a value
+         {response_code, _} <- to_int(raw_response_code),
+         response_body <- try_encode(message) do
```

Comment:
wouldn't it make more sense to not even touch the body in binary mode? I mean, best would be to forward it verbatim - no decoding or encoding

Reply:
So, I've tried and it worked fine with Kafka, but not with Phoenix as it parsed it regardless of format I've tried.

Reply:
we had the same issue with the inbound gateway, where we had to disable Phoenix' parsing of the body entirely. So far we didn't need to do that on rig_api but perhaps it'd make sense for this use case... since I know from experience that this breaks all the other endpoints and is quite a nasty thing to do, I suggest you keep it like this for now.